src/core/store/http/module.js

```diff
@@ -6,11 +6,23 @@ import {
   SyncError, createSyncReducer,
 } from './sync';
 
+function tokenUpdater(token) {
+  localStorage.setItem('token', JSON.stringify(token));
+};
+
+let client = null;
+
 const farm = () => {
-  const host = localStorage.getItem('host');
-  const user = localStorage.getItem('username');
-  const password = localStorage.getItem('password');
-  return farmOS(host, user, password);
+  if (!client) {
+    const host = localStorage.getItem('host');
+    const token = localStorage.getItem('token');
+    client = farmOS(host, 'farm_client', tokenUpdater);
+    if (token == null) {
+      throw new Error('farm not authorized');
+    }
+    client.useToken(JSON.parse(token));
+  }
+  return client;
 };
```

Comment:
@jgaehring the reason we need to instantiate one `client` is so that when a batch of requests is sent from FK they are handled together. This ensures that the `refresh` flow is only attempted once (a `refresh_token` can only be used once), and that subsequent requests will use the new `access_token` that is retrieved. It makes sense to keep this logic in farmOS.js, rather than FK, as this same problem would need to be solved by other users of the library.

> I'm still trying to grok the new internals of farmOS.js, particularly how it refreshes the token, and how that does or does not update the internal farm.token state (does it?). 

Yeah! The internal `farm.token` is managed by the client. When a refresh happens, `farm.token` is always updated. This is when the optional `tokenUpdater` callback is called - for use cases that need to maintain the user's authorization for long amounts of time, they need to store new `tokens` outside of the client.

This is optional because if you were using farmOS.js within a script, you might be fine authorizing the client (with username + password) every time you run the script. If the script runs for a long time, then it is important that the client can refresh internally, even though the script doesn't necessarily need to know about the new token.

> That makes me wonder, in fact, if the token should be an (optional) function parameter passed to farmOS().

We could certainly do that! I opted not to do this because internally we would just be calling the `.useToken()` method (or writing more code with the same logic). Making the user call `.useToken()` is just more explicit - and it's possible that this method could throw an error if the token object is incorrect in any way; a token could have already expired, or the token object could be missing certain attributes. None of that is handled right now, but might be better handled if we require the explicit call to `.useToken()`?

> but perhaps we could change the function signature to accept only 2 parameters: first, the host url, since that is the only required param; second, an options object, which can contain the clientId, tokenUpdater and token as properties

Regardless if we include `token` as an optional parameter, I agree this function signature would be better!

Reply:
> the reason we need to instantiate one client is so that when a batch of requests is sent from FK they are handled together. 

I'm still wondering, however, why do we need to instantiate `client;` then instantiate `farm` as a function which reassigns `client`, when we could just instantiate `farm` as the client itself, and get rid of the unnecessary `client` variable?

Reply:
I see what you're saying. I think you're correct that `client` is unnecessary. But... either way I just realized that there is a bug with this implementation.

If you try to "Sync All Logs" *before* logging in to FK (it fails as expected), then login, and later try to "Sync All Logs" an error is thrown. This is because the farmOS client was not provided a token when it was instantiated (because the login/authorization process had not yet happened)

Overall we just need to refactor how the state of the farmOS client is handled here

Reply:
I think this is exactly why I wrapped the client in a function to begin with (ie, made it "lazy"), but yea, outcome is the same, as you said:

> we just need to refactor how the state of the farmOS client is handled here

Lets see how that shakes out and circle back round to this.